### DISCUSSION TITLE
Proper-Named: Attribute vs Class

### DISCUSSION THREAD

**Discussion by tajmone** [alan-if/alan-i18n#54] [Libraries Changes]
This thread discusses how the library should handle proper-named instances (i.e. taking no `Article`), where the default approach from ALAN Lib v0.6.x was to use a dedicated `actor` subclass (`named_actor`) and the new Spanish Foundation has introduced a dedicated attribute (on `entity`) for the same purpose.

# Spanish Library

**Spanish Foundation** [v0.3.3] introduced in [`gramática.i`][gramática.i] the new `nombre_propio` attribute (defined on `entity`):

> A new `nombre_propio` attribute has been added to handle entities with a proper name (defaults to `Has not`).
> Proper-named entities are given empty strings as their `Definite` and `Indefinite Article` at initialization time, so that no articles are printed when mentioned.

But in [`atributos.i`][atributos.i] we can still find still the definition of the `named_actor` class, left over from the original English **ALAN Lib 0.6** and **pALANte**:

```alan
Every named_actor IsA actor
  Definite Article ""
  Indefinite Article ""
End every named_actor.
```

@Rich15, We need to decide if this specialized class needs to be kept, since we now have also the `nombre_propio` attribute, or whether the attribute should be dropped in favour of the class — in the latter case, we'd have to tweak the initialization code again, reverting to the previous state and use the class instead.

Since Spanish and Italian have to deal with gender and number (GNA) more than the English library, it might make more sense to keep the attribute and drop the class.

But before deciding it might be worth considering the question in more general terms, as this decision might also apply to the English library and all other future locales...

# Proper-Named via Class

From what I gather, the original choice of implementing proper-named via a specialized `actor` subclass originated from the assumption that this feature is only needed for NPCs.

Although this assumption might seem reasonable and obvious, at first glance, it's not only actors which can have proper names in an adventure, but other classes too. It only requires some lateral-thinking brainstorming to realize that in IF adventures we could be many examples of non-actor instances having proper names — a child Hero or NPC might have a teddy bear which is referred to in the game via a proper name (e.g. 'Teddy'), hence needing to drop any `Article`.

And I'm sure that we could come up with many more practical examples involving `object` instances — and even `location`s, e.g. in adventures with verbs that can refer to distant locations (e.g. `travel to (location)!`), where locations might have proper names (e.g. cities, countries, etc.).

Furthermore, note every "game actor" is implemented as an `actor` instance in the code — e.g. when a character dies it's often replaced with an `object` instance, which keeps its previous appearances, including being addressed by a proper name. There are many cases were an author might opt to implement an actor via the `object` class, e.g. a small pet that needs to be carried in containers. The need for `actor` is only really useful to discern which verbs apply to what, or when the instance needs scripting — the latter being also a good reason to implement some non-living game entities as actors, to allow scripting them (e.g. an automaton, an object with a "life of its own", etc.).

So, this might be an argument against the adoption of a specialized sub-class, since being proper-named is not a property exclusive to actors, but can apply to any class, and could be applied to `instance` — which of course means an attribute, since there's no point in sub-classing `instance`.

# Proper-Named via Attribute

It's also worth mentioning that the `named_actor` class was dropped in later incarnations of the ALAN Library, to be replaced by an attribute (e.g. in **ALAN StdLib v2**) — a choice which I believe was motivated from the same (or very similar) practical considerations.

# English Lib vs Spanish/Italian

A final point worth mentioning in this respect, when it comes to deciding on the issue as Foundation Library design choice affecting all libraries, is that the Spanish and Italian libraries rely on a grammar module for initializing gender and number of all game instances, whereas the English library doesn't.

The English library simply takes care of `Article` and `Pronoun` at the specialized classes level, since these are only relevant on `person` and its subclasses — `male` taking "him", and `female` taking "her".

It would be tempting to dismiss the whole argument as an example of a language-specific feature, where diverting from the English Foundation paradigm is justified by the special needs and challenges of the Spanish and Italian languages, keeping the base design to implement proper-naming using classes, and allowing Spanish and Italian to exceptionally use attributes due to GNA complexity.

But I'm not convinced this a valid enough argument for keeping the `named_actor` class — on the contrary, it seems to defeat the whole purpose of having the `male` and `female` classes in the first place, since authors would still have to re-define the correct `Pronoun` on `named_actor` instances.

Relying on a `proper-named` attribute would simplify things, since instances of any class could make use of it (not just `male` and `female`).

Furthermore, the fact that the English grammar is less demanding when it comes to GNA doesn't mean that the English library couldn't benefit from an `Initialize` block — in fact, the **StdLib 2** introduced a library initialization code, to handle various chores on the author's behalf and simplify the task of creating adventures.

# Final Thoughts

Before taking any decisive course of action on this topic (for the Spanish and Italian libraries, or for the whole Foundation design), we might want to discuss it together.

Currently, we have the Spanish library diverging from the English reference library on this, for it introduced the `nombre_propio` attribute to handle this — which seems the most natural approach, given the language needs, and I believe also inspired by the Italian grammar module in this respect.

I personally would like to adopt the same approach for the Italian library, and drop the `named_actor` class altogether (which I believe would make sense in the Spanish lib too, but that's for @Rich15 to decide).

But since there's also the upcoming Swedish library in the works (in its dev branch), I'd like to hear what @thoni56 has to say on the topic, and what solution would work best in Swedish.

As for the English library, I personally think it makes more sense to drop the class in favour of the attribute therein too — but I might be biased on this, due to my prolonged exposure to the ALAN StdLib, where the `proper_named` attribute has replaced the old class.

What I'd like to avoid is the proliferation of diverging library designs across the various ports, so whatever choice we adopt it's good as long as it's a motivated and clear one.

I'd like also to mention here the newly added [`PORTS_STATUS.md`][PORTS_STATUS.md] document, which I introduced to help tracking new features and design changes in the English reference library, and their status in the ported libraries. This should be helpful when new ports are ready to join the `main` branch — e.g. for the upcoming Swedish library, which might need to catch up with the various changes added in the meantime.

[v0.3.3]: https://github.com/alan-if/alan-i18n/blob/main/alan_es/Foundation/CHANGELOG.md#proper-named-actors

[gramática.i]: https://github.com/alan-if/alan-i18n/blob/6d1f3fbd7229f6/alan_es/Foundation/gram%C3%A1tica.i#L20 "View source from 'Foundation ES v0.3.3' commit"
[atributos.i]: https://github.com/alan-if/alan-i18n/blob/6d1f3fbd7229f6/alan_es/Foundation/atributos.i#L18 "View source from 'Foundation ES v0.3.3' commit"

[PORTS_STATUS.md]: https://github.com/alan-if/alan-i18n/blob/main/PORTS_STATUS.md

**Comment by Rich15** [comment]
> Spanish Foundation v0.3.3 introduced in `gramática.i` the new `nombre_propio` attribute (defined on `entity`):
>
> > A new `nombre_propio` attribute has been added to handle entities with a proper name (defaults to `Has not`).
Proper-named entities are given empty strings as their `Definite` and `Indefinite Article` at initialization time, so that no articles are printed when mentioned.
> 
> But in `atributos.i` we can still find still the definition of the `named_actor` class, left over from the original English ALAN Lib 0.6 and pALANte:
> ```
> Every named_actor IsA actor
>   Definite Article ""
>   Indefinite Article ""
> End every named_actor.
> ```
> @Rich15, We need to decide if this specialized class needs to be kept, since we now have also the `nombre_propio` attribute, or whether the attribute should be dropped in favour of the class — in the latter case, we'd have to tweak the initialization code again, reverting to the previous state and use the class instead.
> ...
> I personally would like to adopt the same approach for the Italian library, and drop the named_actor class altogether (which I believe would make sense in the Spanish lib too, but that's for @Rich15 to decide).

I think we should drop the class in favour of the attribute in both Spanish and Italian Libraries. It provides more flexibility, since we could also define a `nombre_propio` to non-actors who need it.

This could also apply to the English Library if it makes things eaiser, but I don't know if it's actually needed.

> I'd like also to mention here the newly added `PORTS_STATUS.md` document, which I introduced to help tracking new features and design changes in the English reference library, and their status in the ported libraries. This should be helpful when new ports are ready to join the `main` branch — e.g. for the upcoming Swedish library, which might need to catch up with the various changes added in the meantime.

Great!

**Reply by tajmone** [reply]
> I think we should drop the class in favour of the attribute in both Spanish and Italian Libraries

I totally agree!

> This could also apply to the English Library if it makes things eaiser, but I don't know if it's actually needed.

Let's put it this way: in the rare case it's needed there would be no way around using the class, except implementing a redundant attribute. In this particular use case, it seems more idiomatic to me to use an attribute instead of a class. But of course, I might be forgetting something that didn't pass through my mind in this respect.
